armory/instrument/config.py

```diff
@@ -0,0 +1,454 @@
+                )
+            total_time = entry["total_time"]
+            execution_count = entry["execution_count"]
+            average_time = total_time / execution_count
+            results[
+                f"Avg. CPU time (s) for {execution_count} executions of {name}"
+            ] = average_time
+            if "stats" in entry:
+                results[f"{name} profiler stats"] = entry["stats"]
+        return results
+
+    def _sink(self, results_dict):
+        """
+        sink for results_writer to write to
+        """
+        self.metric_results = results_dict
+
+    def _metric_results(self):
+        get_hub().close()
+        if self.metric_results is None:
+            log.warning("No metric results received from ResultsWriter")
+            return {}
+        return self.metric_results
+
+    def results(self):
+        results = {}
+        results.update(self._computational_results())
+        results.update(self._metric_results())
+        return results
+
+
+def perturbation_metrics(names, use_mean=True, record_final_only=True):
+    if use_mean:
+        final = np.mean
+    else:
+        final = None
+
+    hub = get_hub()
+    for name in names:
+        metric = metrics.get_supported_metric(name)
+        hub.connect_meter(
+            Meter(
+                f"perturbation_{name}",
+                metric,
+                "scenario.x",
+                "scenario.x_adv",
+                final=final,
+                final_name=f"perturbation_mean_{name}",
+                record_final_only=record_final_only,
+            )
+        )
+
+
+MEAN_AP_METRICS = [
+    "object_detection_AP_per_class",
+    "apricot_patch_targeted_AP_per_class",
+    "dapricot_patch_targeted_AP_per_class",
+    "carla_od_AP_per_class",
+]
+# quanity_metrics only impacts output printing
+QUANTITY_METRICS = [
+    "object_detection_hallucinations_per_image",
+    "carla_od_hallucinations_per_image",
+]
+
+
+class ResultsLogWriter(LogWriter):
+    """
+    Logs successful results (designed for task metrics)
+    """
+
+    def __init__(
+        self,
+        adversarial=False,
+        targeted=False,
+        used_preds_as_labels=False,
+        log_level: str = "SUCCESS",
+    ):
+        super().__init__(log_level=log_level)
+        if targeted:
+            if adversarial:
+                self.wrt = "target"
+                self.task_type = "adversarial"
+            else:
+                raise ValueError("benign task cannot be targeted")
+        elif adversarial:
+            if used_preds_as_labels:
+                self.wrt = "benign predictions as"
+            else:
+                self.wrt = "ground truth"
+            self.task_type = "adversarial"
+        else:
+            self.wrt = "ground truth"
+            self.task_type = "benign"
+
+    def _write(self, name, batch, result):
+        # TODO: once metrics have also been updated, rewrite to be less error prone
```

Comment:
do you intend to address this TODO before merge?

Reply:
no, this is meant to be addressed after updating `armory.utils.metrics` in a future PR (post 0.15.0)